pkg/controller/controller.go

```diff
@@ -0,0 +1,200 @@
+	NeedsProcessing(resource R, logger log.FieldLogger, eventRecorder record.EventRecorder) bool
+}
+
+func (r *reconciler[R]) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.logger.WithFields(log.Fields{
+		"namespace": req.Namespace,
+		"name":      req.Name,
+	})
+
+	logger.Infof("Reconciling")
+	defer func() {
+		logger.Infof("Finished reconciling")
+	}()
+
+	fail := func(err error, requeue bool) (ctrl.Result, error) {
+		logger.Error(err)
+		cr := &ctrl.Result{}
+		if requeue {
+			cr.RequeueAfter = r.requeueInterval
+		}
+		return *cr, nil
+	}
+
+	resource := r.inner.New()
+
+	err := r.Get(ctx, req.NamespacedName, resource)
+	switch {
+	case errors.IsNotFound(err):
+		return fail(fmt.Errorf("resource deleted from cluster; noop"), false)
+	case err != nil:
+		return fail(fmt.Errorf("unable to retrieve resource from cluster: %s", err), true)
+	}
+
+	logger = logger.WithFields(r.inner.LogDetail(resource))
+
+	status := resource.GetStatus()
+	status.UpdateObservationData(resource.GetGeneration())
+	if status.SynchronizationTime != nil {
+		logger.Infof("Last synchronization time: %s", status.SynchronizationTime)
+	} else {
+		logger.Info("Resource not synchronized before")
+	}
+
+	if resource.GetDeletionTimestamp() != nil {
+		logger.Info("Resource is marked for deletion")
+		err = r.inner.Delete(resource, logger, r.eventRecorder)
+		if err != nil {
+			return fail(fmt.Errorf("failed to delete resource: %w", err), true)
+		}
+		if controllerutil.RemoveFinalizer(resource, r.finalizer) {
+			err = r.Update(ctx, resource)
+			if err != nil {
+				return fail(fmt.Errorf("failed to remove finalizer: %w", err), true)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	err = resource.ApplyDefaults()
+	if err != nil {
+		return fail(err, false)
+	}
+
+	hash, err := resource.Hash()
+	if err != nil {
+		return fail(err, false)
+	}
+
+	if !r.needsProcessing(resource, hash, logger) {
+		return ctrl.Result{}, nil
+	}
+
+	result := r.inner.Process(resource, logger, r.eventRecorder)
+
+	status.UpdateSynchronizationData(hash, result.State, resource.GetGeneration())
+	conditions := append(result.Conditions, r.calculateSucess(result.Conditions, result.State))
+	status.UpdateConditions(conditions, resource.GetGeneration())
+
+	if controllerutil.AddFinalizer(resource, r.finalizer) {
+		err = r.Update(ctx, resource)
+	} else {
+		err = r.Status().Update(ctx, resource)
+	}
+	if err != nil {
+		return fail(err, true)
+	}
+
+	if result.State == SynchronizationStateFailed {
+		return fail(fmt.Errorf("synchronisation failed"), result.Requeue)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *reconciler[R]) needsProcessing(resource R, hash string, logger log.FieldLogger) bool {
+	status := resource.GetStatus()
+	if status.SynchronizationHash != hash {
+		logger.Infof("Resource has changed since last synchronization: %s != %s", status.SynchronizationHash, hash)
+		return true
+	}
+
```

Comment:
Could this just be a helper function insted, so that hash-es aren't a requirement?

Reply:
Afaik, all our operators use the hash check procedure today, but I also believe we might want to stop doing that at some point. I'll explore how it could look with a helper function ...

I guess one way to do it would be that the `r.inner.NeedsProcessing` call would get a function that does this check, and then be able to opt-out of using it ... :thinking: